File: app/schemas/device.py
from pydantic import BaseModel, Field, IPvAnyAddress
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from enum import Enum

class DeviceType(str, Enum):
    STANDARD = "standard"
    CUSTOM = "custom"

class OSType(str, Enum):
    WINDOWS = "windows"
    MACOS = "macos"
    LINUX = "linux"

# Base schemas
class DeviceBase(BaseModel):
    name: str
    description: Optional[str] = None
    ip_address: IPvAnyAddress

class StandardDeviceBase(BaseModel):
    os_type: OSType
    hostname: str


class CustomDeviceBase(BaseModel):
    plugin_id: int
    connection_params: Dict[str, Any] = {}

# Create schemas
class DeviceCreate(DeviceBase):
    type: DeviceType
    standard_device: Optional[StandardDeviceBase] = None
    custom_device: Optional[CustomDeviceBase] = None

# Update schemas
class DeviceUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    ip_address: Optional[IPvAnyAddress] = None
    is_active: Optional[bool] = None

class StandardDeviceUpdate(BaseModel):
    os_type: Optional[OSType] = None
    hostname: Optional[str] = None

class CustomDeviceUpdate(BaseModel):
    plugin_id: Optional[int] = None
    connection_params: Optional[Dict[str, Any]] = None

# Response schemas
class StandardDeviceResponse(StandardDeviceBase):
    id: int

    class Config:
        from_attributes = True

class CustomDeviceResponse(CustomDeviceBase):
    id: int
    plugin_name: Optional[str] = None

    class Config:
        from_attributes = True

class DeviceResponse(DeviceBase):
    id: int
    type: DeviceType
    is_active: bool
    created_at: datetime
    updated_at: datetime
    standard_device: Optional[StandardDeviceResponse] = None
    custom_device: Optional[CustomDeviceResponse] = None

    class Config:
        from_attributes = True

File: infra-platform-frontend/src/types/device.ts
// src/types/device.ts
export type DeviceType = 'standard' | 'custom';
export type OSType = 'windows' | 'macos' | 'linux';

export interface StandardDeviceData {
  id: number;
  os_type: OSType;
  hostname: string;
}

export interface CustomDeviceData {
  id: number;
  plugin_id: number;
  plugin_name: string;
  connection_params: Record<string, any>;
}

export interface Device {
  id: number;
  name: string;
  description: string | null;
  ip_address: string;
  type: DeviceType;
  created_at: string;
  updated_at: string;
  standard_device: StandardDeviceData | null;
  custom_device: CustomDeviceData | null;
}

export interface StandardDeviceCreate {
  os_type: OSType;
  hostname: string;
}

export interface CustomDeviceCreate {
  plugin_id: number;
  connection_params: Record<string, any>;
}

export interface DeviceCreate {
  name: string;
  description?: string;
  ip_address: string;
  type: DeviceType;
  standard_device?: StandardDeviceCreate;
  custom_device?: CustomDeviceCreate;
}

export interface DeviceUpdate {
  name?: string;
  description?: string;
  ip_address?: string;
}

export interface DeviceStats {
  cpu?: {
    usage: number;
    cores: { core: number; usage: number }[];
    temperature: any[];
  };
  memory?: {
    total: number;
    used: number;
    free: number;
    percent: number;
  };
  disk?: {
    device: string;
    mountpoint: string;
    total: number;
    used: number;
    free: number;
    percent: number;
  }[];
  network?: {
    interface: string;
    rx: number;
    tx: number;
    rx_packets: number;
    tx_packets: number;
  }[];
  processes?: {
    total: number;
    running: number;
    sleeping: number;
    thread: number;
  };
  system?: {
    hostname: string;
    os_name: string;
    os_version: string;
    uptime: string;
  };
  error?: string;
}